RFCs/FS-1152-For-loop-with-accumulation.md

```diff
@@ -0,0 +1,143 @@
+    <expr>
+```
+The value of the loop body is used to update the accumulator which is returned in the end.
+
+# Motivation
+
+Currently, to do this, we need an additional mutable variable. For example,
+```fs
+let mutable sentence = "" // mutable accumulator
+for word in ["Hello", " ", "World", "!"] do
+    sentence <- sentence + word
+printfn "%s" sentence
+```
+
+However, the mutable variable leaks outside the loop body, which is undesirable.
+Moreover, the accumulator must be mutable - which is against functional immutable semantics.
+
+It is proposed that the accumulator be embeddable into the loop itself:
+```fs
+for sentence = "" with word in ["Hello", " ", "World", "!"] do
+    sentence + word
+|> printfn "%s"
+```
+The return value of the loop body updates the accumulator.
+This offers a desirable middle ground between `for` loops that must only have side-effects (returning `unit`)
+and purely functional `fold`s that must be in lambda form.
+
+Meanwhile, folds are hard to understand and the body cannot be use computation expression functionality.
+```fs
+["Hello", " ", "World", "!"]
+|> Seq.fold (fun sentence -> sentence + word (*loop body, and yet we lose the ability to do let!*)) "" // initial state is placed last??
+|> printfn "%s"
+```
+which suggests a missed opportunity to make them more familiar to people who know `for` loops.
+
+# Detailed design
+
+For this syntax
+
+```fs
+for <pat1> = <expr1> with <pat2> in <expr2> do
+    <exprs>
+    <expr3>
+```
+
+it undergoes a simple syntactical expansion to
+
+```fs
+let mutable <accum> = <expr1>
+for <pat2> in <expr2> do
+    let <pat1> = <accum>
+    <exprs>
+    <accum> <- <expr3>
+<accum>
+```
+where `<accum>` is a compiler-generated accumulation variable that cannot be used outside the loop,
+and `<exprs>` match any number of expressions before the final expression in the loop body.
+
+This means that `<accum>` is the value of this loop. It is also the accumulator which gets updated each iteration.
+
+Note that `<exprs>` are subject to usual treatment by computation expressions including application of implicit yields,
+which is different from the final expression `<expr3>`.
+
+Therefore, this may happen:
+```fs
+let result = [
+    for sentence = "" with word in ["Hello", " ", "World", "!"] do
+        sentence
+        sentence + word
+    |> printfn "%s"
+]
+// expands to
+let result = [
+    let mutable <accum> = ""
+    for word in ["Hello", " ", "World", "!"] do
+        sentence // Implicit yield here
+        <accum> <- sentence + word // But not for the final expression of the loop body
+    <accum>
+    |> printfn "%s"
+]
+```
+
+# Drawbacks
+
+It's another syntax to learn.
+
+# Alternatives
+
+Doing nothing - we miss an opportunity to bridge regular for loops with lambda folds.
+
+There is a potential extension to folding across multiple sequences with the `and` keyword in the future -
+```fs
+for s, p = 0, 0 with t in ts and u in us do
+    s + t, p + u
+```
```

Comment:
This section should mention computation expressions: https://github.com/fsharp/fslang-suggestions/issues/1362#issuecomment-2123061400.